presto-main/src/main/java/com/facebook/presto/execution/buffer/SpoolingOutputBuffer.java

```diff
@@ -0,0 +1,788 @@
+import static com.facebook.presto.spi.page.PagesSerdeUtil.readSerializedPages;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.Iterators.advance;
+import static com.google.common.util.concurrent.Futures.catchingAsync;
+import static com.google.common.util.concurrent.Futures.immediateFailedFuture;
+import static com.google.common.util.concurrent.Futures.immediateFuture;
+import static com.google.common.util.concurrent.Futures.transform;
+import static com.google.common.util.concurrent.Futures.transformAsync;
+import static com.google.common.util.concurrent.MoreExecutors.listeningDecorator;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newCachedThreadPool;
+
+@SuppressWarnings("UnstableApiUsage")
+public class SpoolingOutputBuffer
+        implements OutputBuffer
+{
+    private final OutputBuffers outputBuffers;
+    private final StateMachine<BufferState> state;
+    private final String taskInstanceId;
+    private final TempDataOperationContext context;
+    private final TempStorage tempStorage;
+    private final DataSize threshold;
+    private final ListeningExecutorService executor;
+
+    private final AtomicLong totalBufferedBytes = new AtomicLong();
+    private final AtomicLong totalBufferedPages = new AtomicLong();
+    private final AtomicLong totalPagesAdded = new AtomicLong();
+    private final AtomicLong totalRowsAdded = new AtomicLong();
+
+    private final OutputBufferId outputBufferId = new OutputBufferId(0);
+    private final Closer closer = Closer.create();
+
+    private static ExecutorService coreExecutor = newCachedThreadPool(daemonThreadsNamed("spooling-outputbuffer-%s"));
+    private static final Logger log = Logger.get(SpoolingOutputBuffer.class);
+
+    private final AtomicBoolean noMorePages = new AtomicBoolean();
+    private final AtomicLong currentMemorySequenceId = new AtomicLong();
+    private final AtomicLong currentSequenceId = new AtomicLong();
+    private final AtomicLong startPage = new AtomicLong();
+    private final AtomicLong totalPagesRemaining = new AtomicLong();
+    private final AtomicLong totalInMemoryBytes = new AtomicLong();
+    private final AtomicLong peakMemoryUsage = new AtomicLong();
+
+    private final AtomicLong totalStorageBytesAdded = new AtomicLong();
+    private final AtomicLong totalStoragePagesAdded = new AtomicLong();
+
+    @GuardedBy("this")
+    private final LinkedList<HandleInfo> handleInfoQueue = new LinkedList<>();
+
+    @GuardedBy("this")
+    private final Queue<SerializedPage> pages = new ArrayDeque<>();
+
+    @GuardedBy("this")
+    private PendingRead pendingRead;
+
+    private FinalizerService finalizerService;
```

Comment:
```suggestion
    private final FinalizerService finalizerService;
```

Also move this to the other initialized arguments.